File: python_sdk/src/reality_capture/service/estimation.py
from pydantic import BaseModel, Field
from reality_capture.service.job import JobType, _get_appropriate_service, Service
from typing import Union
from enum import Enum
from reality_capture.specifications.calibration import CalibrationSpecificationsCreate, CalibrationCost
from reality_capture.specifications.change_detection import ChangeDetectionSpecificationsCreate
from reality_capture.specifications.constraints import ConstraintsSpecificationsCreate, ConstraintsCost
from reality_capture.specifications.fill_image_properties import (FillImagePropertiesSpecificationsCreate,
                                                                  FillImagePropertiesCost)
from reality_capture.specifications.import_point_cloud import ImportPCSpecificationsCreate, ImportPCCost
from reality_capture.specifications.objects2d import Objects2DSpecificationsCreate
from reality_capture.specifications.production import ProductionSpecificationsCreate, ProductionCost
from reality_capture.specifications.reconstruction import ReconstructionSpecificationsCreate, ReconstructionCost
from reality_capture.specifications.segmentation2d import Segmentation2DSpecificationsCreate
from reality_capture.specifications.segmentation3d import Segmentation3DSpecificationsCreate
from reality_capture.specifications.segmentation_orthophoto import SegmentationOrthophotoSpecificationsCreate
from reality_capture.specifications.tiling import TilingSpecificationsCreate, TilingCost
from reality_capture.specifications.touchup import (TouchUpImportSpecificationsCreate,
                                                    TouchUpExportSpecificationsCreate,
                                                    TouchUpExportCost, TouchUpImportCost)
from reality_capture.specifications.water_constraints import WaterConstraintsSpecificationsCreate, WaterConstraintsCost


class CostEstimationCreate(BaseModel):
    type: JobType = Field(description="Type of job.")
    specifications: Union[CalibrationSpecificationsCreate, ChangeDetectionSpecificationsCreate,
                          ConstraintsSpecificationsCreate,
                          FillImagePropertiesSpecificationsCreate, ImportPCSpecificationsCreate,
                          Objects2DSpecificationsCreate, ProductionSpecificationsCreate,
                          ReconstructionSpecificationsCreate, Segmentation2DSpecificationsCreate,
                          Segmentation3DSpecificationsCreate, SegmentationOrthophotoSpecificationsCreate,
                          TilingSpecificationsCreate, TouchUpExportSpecificationsCreate,
                          TouchUpImportSpecificationsCreate, WaterConstraintsSpecificationsCreate] = (
        Field(description="Specifications aligned with the job type."))
    cost_parameters: Union[CalibrationCost, ConstraintsCost, FillImagePropertiesCost,
                           ImportPCCost, ProductionCost, ReconstructionCost, TilingCost,
                           TouchUpExportCost, TouchUpImportCost, WaterConstraintsCost] = (
        Field(description="Cost format aligned with the job type.", alias="costParameters"))

    def get_appropriate_service(self) -> Service:
        """
        Return the appropriate service for such a job.

        :return: Service enum.
        """
        return _get_appropriate_service(self.type)


class UnitType(Enum):
    MODELING = "Modeling"
    ANALYSIS = "Analysis"
    CONVERSION = "Conversion"


class CostEstimation(CostEstimationCreate):
    unique_id: str = Field(description="Estimation id", alias="id")
    estimated_units: float = Field(description="Estimated number of units that the job will cost.",
                                   alias="estimatedUnits")
    unit_type: UnitType = Field(description="Type of unit for the job.", alias="unitType")

File: typescript_sdk/src/reality_capture/service/estimation.ts
import { z } from "zod";
import { JobType, getAppropriateService, Service } from "../job";
import { CalibrationSpecificationsCreate, CalibrationCost } from "../../specifications/calibration";
import { ChangeDetectionSpecificationsCreate } from "../../specifications/change_detection";
import { ConstraintsSpecificationsCreate, ConstraintsCost } from "../../specifications/constraints";
import { FillImagePropertiesSpecificationsCreate, FillImagePropertiesCost } from "../../specifications/fill_image_properties";
import { ImportPCSpecificationsCreate, ImportPCCost } from "../../specifications/import_point_cloud";
import { Objects2DSpecificationsCreate } from "../../specifications/objects2d";
import { ProductionSpecificationsCreate, ProductionCost } from "../../specifications/production";
import { ReconstructionSpecificationsCreate, ReconstructionCost } from "../../specifications/reconstruction";
import { Segmentation2DSpecificationsCreate } from "../../specifications/segmentation2d";
import { Segmentation3DSpecificationsCreate } from "../../specifications/segmentation3d";
import { SegmentationOrthophotoSpecificationsCreate } from "../../specifications/segmentation_orthophoto";
import { TilingSpecificationsCreate, TilingCost } from "../../specifications/tiling";
import { TouchUpImportSpecificationsCreate, TouchUpExportSpecificationsCreate, TouchUpExportCost, TouchUpImportCost } from "../../specifications/touchup";
import { WaterConstraintsSpecificationsCreate, WaterConstraintsCost } from "../../specifications/water_constraints";

export const CostEstimationCreateSchema = z.object({
  type: JobType.describe("Type of job."),
  specifications: z.union([
    CalibrationSpecificationsCreate,
    ChangeDetectionSpecificationsCreate,
    ConstraintsSpecificationsCreate,
    FillImagePropertiesSpecificationsCreate,
    ImportPCSpecificationsCreate,
    Objects2DSpecificationsCreate,
    ProductionSpecificationsCreate,
    ReconstructionSpecificationsCreate,
    Segmentation2DSpecificationsCreate,
    Segmentation3DSpecificationsCreate,
    SegmentationOrthophotoSpecificationsCreate,
    TilingSpecificationsCreate,
    TouchUpExportSpecificationsCreate,
    TouchUpImportSpecificationsCreate,
    WaterConstraintsSpecificationsCreate,
  ]).describe("Specifications aligned with the job type."),
  costParameters: z.union([
    CalibrationCost,
    ConstraintsCost,
    FillImagePropertiesCost,
    ImportPCCost,
    ProductionCost,
    ReconstructionCost,
    TilingCost,
    TouchUpExportCost,
    TouchUpImportCost,
    WaterConstraintsCost,
  ]).describe("Cost format aligned with the job type.")
});
export type CostEstimationCreate = z.infer<typeof CostEstimationCreateSchema>;

export function getAppropriateServiceForEstimation(estimation: CostEstimationCreate): Service {
  return getAppropriateService(estimation.type);
}

export enum UnitType {
  MODELING = "Modeling",
  ANALYSIS = "Analysis",
  CONVERION = "Conversion",
}

export const CostEstimationSchema = CostEstimationCreateSchema.extend({
  id: z.string().describe("Estimation id"),
  estimatedUnits: z.number().describe("Estimated number of units that the job will cost."),
  unitType: z.nativeEnum(UnitType).describe("Type of unit for the job."),
});

export type CostEstimation = z.infer<typeof CostEstimationSchema>;